chain/epoch_manager/src/lib.rs

```diff
@@ -1050,12 +1075,75 @@ impl EpochManager {
         }
         Ok(*self.epoch_id_to_start.cache_get(epoch_id).unwrap())
     }
+
+    /// Get epoch info aggregator and update it to block info as of `last_block_hash`. If `epoch_id`
+    /// doesn't match the epoch id of the existing aggregator, re-initialize the aggregator.
+    /// If `copy_only` is true, then we clone what is in the cache. Otherwise we take the aggregator
+    /// from cache and invalidates the cache.
+    pub fn get_and_update_epoch_info_aggregator(
+        &mut self,
+        epoch_id: &EpochId,
+        last_block_hash: &CryptoHash,
+        copy_only: bool,
+    ) -> Result<EpochInfoAggregator, EpochError> {
+        let epoch_info_aggregator_cache = if copy_only {
+            self.epoch_info_aggregator.clone()
+        } else {
+            self.epoch_info_aggregator.take()
+        };
+        let mut aggregator = if let Some(aggregator) = epoch_info_aggregator_cache {
+            aggregator
+        } else {
+            self.store
+                .get_ser(ColEpochInfo, AGGREGATOR_KEY)
+                .map_err(EpochError::from)?
+                .unwrap_or_else(|| EpochInfoAggregator::new(epoch_id.clone(), *last_block_hash))
+        };
+        let mut epoch_change = false;
+        if &aggregator.epoch_id != epoch_id {
+            aggregator = EpochInfoAggregator::new(epoch_id.clone(), *last_block_hash);
+            epoch_change = true;
+        }
+        let epoch_info = self.get_epoch_info(epoch_id)?.clone();
+        let mut proposals: HashMap<AccountId, ValidatorStake> = HashMap::default();
+        let mut cur_hash = *last_block_hash;
+        while cur_hash != aggregator.last_block_hash || epoch_change {
+            // Avoid cloning
+            let prev_hash = self.get_block_info(&cur_hash)?.prev_hash;
+            let prev_height = self.get_block_info(&prev_hash)?.height;
+            let block_info = self.get_block_info(&cur_hash)?;
+            if &block_info.epoch_id != epoch_id || block_info.prev_hash == CryptoHash::default() {
+                break;
+            }
+            aggregator.update_trackers(&block_info, &epoch_info, prev_height);
+
+            for proposal in block_info.proposals.iter() {
+                proposals.entry(proposal.account_id.clone()).or_insert_with(|| proposal.clone());
```

Comment:
Can the same block contain multiple proposals for the same account? If yes, you probably want to do `rev()` on the iterator above so that the latter of the two is kept.

Reply:
No. My change in runtime ensures that for every account, it can have at most one proposal in every chunk (and therefore every block).